### ISSUE TITLE
Explore more obvious section headers in quick pick

### ISSUE THREAD

**Issue by daviddossett** [microsoft/vscode#149057]
The right-aligned label text has always been slightly confusing to me as a pattern to separate sections in quick pick lists. I think it's worth exploring how we might make these more obvious while still using vertical space efficiently.

## Current

<img width="631" alt="CleanShot 2022-05-08 at 21 29 37@2x" src="https://user-images.githubusercontent.com/25163139/167340909-4237b4a9-d6ed-4cb3-849e-4fe016625122.png">

### GitHub

<img width="812" alt="CleanShot 2022-05-08 at 21 30 48@2x" src="https://user-images.githubusercontent.com/25163139/167341026-78dbab4a-8c82-4107-a6cf-147cc2bab906.png">

### Raycast 

<img width="770" alt="CleanShot 2022-05-08 at 21 31 09@2x" src="https://user-images.githubusercontent.com/25163139/167341043-34db2bd9-a15e-42dc-abcd-90ecf9d76ca0.png">

cc @lychung7 @misolori

**Comment by jaeyonglee8687** [comment]
Sharing another example 

**Runner Pro**
https://docs.sketchrunner.com/search-commands/run
<img width="1427" alt="image" src="https://user-images.githubusercontent.com/87626977/168940807-645d5b88-c8b9-4a73-9486-e5a5b97b30f1.png">

**Comment by daviddossett** [comment]
A couple of possible solutions within the current quick pick design:

<img width="780" alt="CleanShot 2022-06-10 at 10 52 15@2x" src="https://user-images.githubusercontent.com/25163139/173123439-60897458-ef36-4729-870e-0730f6fc5a69.png">

<img width="785" alt="CleanShot 2022-06-10 at 10 52 20@2x" src="https://user-images.githubusercontent.com/25163139/173123448-f4a22557-ce4e-45e5-95f8-e97ba9b31bf5.png">

**Comment by daviddossett** [comment]
This could also be extended into an [idea I had](https://twitter.com/david_dossett/status/1535326011985784832) for a "comfortable" density mode for the quick pick/command palette.

<img width="910" alt="CleanShot 2022-05-17 at 20 50" src="https://user-images.githubusercontent.com/25163139/173128144-d7117f0c-d66c-4914-928c-ae8f0bb8cb26.png">

**Comment by daviddossett** [comment]
Extending these explorations to include ideas for actions in the section header rows. Some early ideas:

## With Inline Icon Actions
![with-section-headers](https://user-images.githubusercontent.com/25163139/183467618-3e736012-2251-4ec3-87ba-c72a3a7142e3.gif)

## With Icon Action Labels

We could also use labels here since the inline icons alone would be a downgrade in terms of understanding what the command is—today we use the plain text as a quick pick entry so this could retain some of that clarity. 

![with-action-labels](https://user-images.githubusercontent.com/25163139/183467676-4405a9d6-ce26-4ba1-855f-637c12d39be9.gif)

cc @misolori

**Comment by TylerLeonhardt** [comment]
I like that the separator buttons could have text to them though I'm sure the first feature request once that lands is to allow buttons in the normal quick pick items to have text :) not a bad idea but I am concerned about space in the quickpick...

**Comment by misolori** [comment]
I think starting with icons (no labels) is a good start to match most of our inline actions. I also wonder if we can have the section header actions/toolbar appear when you are on a section, just like views?

<img width="307" alt="CleanShot 2022-08-10 at 10 46 28@2x" src="https://user-images.githubusercontent.com/35271042/183981278-76d923e1-9078-466c-ad0c-be3c3f47b40b.png">

**Comment by daviddossett** [comment]
Agreed.

> I also wonder if we can have the section header actions/toolbar appear when you are on a section, just like views?

By that do you mean like if you're navigating via keyboard we should up the parent actions? 👍 if so!

**Comment by misolori** [comment]
Right now we show the toolbar whenever you are hovering in a view and then it hides when you are off, as opposed to only showing it when you are hovering over the section title.

**Comment by daviddossett** [comment]
Oh ok even better. Great idea.

**Comment by jaeyonglee8687** [comment]
Love the iterations. The blue color title seems to have a function or embedded link it it at the first glance. It worked when it was on the same line with a quickPick item, but it becomes a little confusing with the separation between section title and pickPick items.

**Comment by daviddossett** [comment]
> The blue color title seems to have a function or embedded link it it at the first glance

That's a fair point. Darker or lighter greys can in some ways make it harder to distinguish between sections. Maybe some spacing above each section would help here:

<img width="451" alt="CleanShot 2022-08-11 at 11 06 04@2x" src="https://user-images.githubusercontent.com/25163139/184208741-a48426eb-69f1-4ce9-99cf-13a805e72f6e.png">

**Comment by daviddossett** [comment]
Incoroporated a few bits of feedback and fixes:
- Added dedicated divder rows to help distinguish between sections
- Added @misolori's suggestion to show the actions on hover/focus of children

![CleanShot 2022-08-15 at 09 32 47](https://user-images.githubusercontent.com/25163139/184676546-1ef3248b-d9f9-40de-940e-8eb1f5152a8d.gif)

**Comment by TylerLeonhardt** [comment]
@daviddossett two questions:
* Do you think we should keep the existing look if no separators have buttons?
* If only a single separator has buttons, should they the UI of separators be aligned to all take up an entire row?

**Comment by daviddossett** [comment]
> Do you think we should keep the existing look if no separators have buttons?

My gut feeling is that it should be a global change. There are a bunch of differences in font, spacing, etc. that makes me think it would be confusing for users to see different formats across different lists. Here's how it would look on a regular grouped quick pick list:

<img width="687" alt="CleanShot 2022-08-15 at 10 22 14@2x" src="https://user-images.githubusercontent.com/25163139/184688422-9e640a76-7d1f-4dff-b0b6-613284c2f8c6.png">

Some inspiration from Safari for those worried about keyboard interactions. Feels super natural to arrow up/down across sections:

![CleanShot 2022-08-15 at 10 54 03](https://user-images.githubusercontent.com/25163139/184689203-80e2593d-5ffe-40a9-84ea-69abe5da770a.gif)

**Comment by eamodio** [comment]
I like the latest iteration. Although I think the section headers are too bold -- the blue stands out and draws the eye much more than the actual choices. I would prefer a style closer to the safari example where the headers are more muted, so the focus stays on the choices themselves.

**Comment by daviddossett** [comment]
I like the grey headers too, at least in this particular default theme. I've landed on that for now. We're also looking at how to better integrate section-specific actions. Today they're often placed at the bottom of a long quick pick list and can be easily missed.

## Header w/ Toolbar

Shows up on hover anywhere in that section. Advantages here are the similarities to a view toolbar and reduced noise in the list. Possible makes it difficult to have a natural-feeling keyboard navigation. We also lose the descriptive list item labels + search capability we have today.

![CleanShot 2022-08-18 at 14 38 49](https://user-images.githubusercontent.com/25163139/185499992-1890bfb4-1faa-4ebd-bdd5-724b5a45ad9d.gif)
 
## Header with Pinned Action Items

We could have section-specific actions rendered at the top of the relevant section as normal list items. This could be achieved today but it would obviously need to be done manually. Maybe there's a way to pin items per section?

<img width="800" alt="CleanShot 2022-08-18 at 14 43 33@2x" src="https://user-images.githubusercontent.com/25163139/185500505-25487698-969f-4f3b-a51c-ad09d521ccc5.png">

**Comment by usernamehw** [comment]
Any chance of keeping headers even after typing? For example, Theme Picker with empty input shows:

```
──────────────────────────────────────────────────────────── light themes
...
GitHub Light
GitHub Light Colorblind
Light (Visual Studio)
Light+ (default light)
...
──────────────────────────────────────────────────────────── dark themes
...
Dark (Visual Studio)
Dark+ (default dark)
GitHub Dark
GitHub Dark Colorblind
...
```

After filtering they are no longer grouped (input content - `github`):

```
GitHub Dark
GitHub Light
GitHub Dark Colorblind
GitHub Light Colorblind
...
```

Ideally, groups should persist:

```
──────────────────────────────────────────────────────────── light themes
GitHub Light
GitHub Light Colorblind
──────────────────────────────────────────────────────────── dark themes
GitHub Dark
GitHub Dark Colorblind
```

**Comment by daviddossett** [comment]
@usernamehw I think that's a reasonable proposal but I'd need to see how to approach quick picks that show something different after typing. There are some examples of presenting other content instead of the filtered list.

**Comment by TylerLeonhardt** [comment]
@usernamehw that sounds like a bug to me because I get that rough experience:
<img width="525" alt="image" src="https://user-images.githubusercontent.com/2644648/194915454-4ac07a68-065c-49e3-b64b-7fb078484a2d.png">

Can you open a new issue and we can get that looked at?

**Comment by daviddossett** [comment]
It does look like there is some strangeness going on here.

<img width="645" alt="CleanShot 2022-10-10 at 09 54 22@2x" src="https://user-images.githubusercontent.com/25163139/194917075-dd08cb92-9097-4f5f-9e5f-dc75bf28dc4d.png">

**Comment by TylerLeonhardt** [comment]
@aeschli is this feedback for you? ^

**Comment by aeschli** [comment]
Agree that this looks strange. Filtering & sorting doesn't go with the separator header. 
It's probably a bug with quick fix dialog to still show the separator labels when filtering (and sorting)
Only when `sortByLabel: false` is set the separators can be shown while filtering. That what is done in the remote indicator menu:
![image](https://user-images.githubusercontent.com/6461412/195031065-eb48683a-2444-4d37-9a72-56295865157a.png)

**Comment by usernamehw** [comment]
Is `sortByLabel` available to extension api?
